Scripts/Runtime/Entities/Tasks/Data/MainThreadTaskWorkData.cs

```diff
@@ -0,0 +1,33 @@
+using Anvil.Unity.DOTS.Data;
+using Anvil.Unity.DOTS.Jobs;
+
+namespace Anvil.Unity.DOTS.Entities
+{
+    /// <summary>
+    /// An <see cref="AbstractTaskWorkData"/> specific for use on the main thread.
+    /// </summary>
+    public class MainThreadTaskWorkData : AbstractTaskWorkData
+    {
+        private static readonly int MAIN_THREAD_INDEX = ParallelAccessUtil.CollectionIndexForMainThread();
+        
+        internal MainThreadTaskWorkData(AbstractTaskDriverSystem system) : base(system)
+        {
+        }
+        
+        /// <inheritdoc cref="GetVDUpdater{TKey,TInstance}"/>
+        public sealed override VDUpdater<TKey, TInstance> GetVDUpdater<TKey, TInstance>()
+        {
+            VDUpdater<TKey, TInstance> updater = base.GetVDUpdater<TKey, TInstance>();
+            updater.InitForThread(MAIN_THREAD_INDEX);
+            return updater;
+        }
+
+        /// <inheritdoc cref="GetVDWriter{TKey,TInstance}"/>
+        public sealed override VDWriter<TInstance> GetVDWriter<TKey, TInstance>()
+        {
+            VDWriter<TInstance> writer = base.GetVDWriter<TKey, TInstance>();
+            writer.InitForThread(MAIN_THREAD_INDEX);
+            return writer;
+        }
```

Comment:
> We can't do this because at the time you call this you don't know what thread you're going to be running on.

ah, ok I thought this might be called at the execution phase instead of schedule phase.


> Hmmn we could do that but it feels very unlike you. We'd be doing 
> MainThreadTaskWorkData.GetMainThreadWriter which seems redundant. We already know were operating on a MainThreadTaskWorkData and it's going to give us a VDWriter that is configured for doing main thread work.

Yea sorry, I mashed two ideas together there. If you wrap you wouldn't need to rename.

By wrapping into a completely different type you'd have specialized your existing `AbstractTaskWorkData` so you'd have code that was something like:
``` csharp
m_MyTaskWorkData.OnThread(MAIN_THREAD_INDEX).GetVDWriter()
```

I guess my hang up with the current setup is that we wouldn't expect it to be configuring threads for us. The other type(s) of TaskWorkData don't initialize a thread for you.

---

Alternatively, should we just do like we've done in other classes and have one type that supplies both sync and async methods? Then the GetAsync it's assumed something else will lock to thread.

---

Another thought....although probably too much work to be worth changing at this point... would be to have VDWriter/Updater be a type of factory container that can't actually do anything until you call `.ForThread()` on them. Then you couldn't ever accidentally forget to initialize the writer/updater/readers to a thread and main thread access would be forced to make that configuration on the spot.
`m_MyTaskWorkData.GetWriter().OnThread(MAIN_THREAD_INDEX)`

Reply:
Let's chat on this one.

Reply:
Chatted and agreed on a direction:
 - Move all logic in `JobTaskWorkData` and `MainThreadTaskWorkData` into one type (getting rid of `AbstractTaskWorkData`)
 - For Main thread vs job access we use the `Async` suffix for job specific acquisition.
 - Rename `MAIN_THREAD_INDEX` to something like `SYNC_THREAD_INDEX` (or similar) with a comment to indicate that we don't NEED to use main thread but we had to choose some arbitrary bucket to write into so we may as well write into main.